README.md
#qTunnel

### qTunnel - the tunnel program wrapping communication on getqujing.net

`qtunnel` is a network proxy to work as an encryption wrapper between remote client and local or remote server. It can work as a stunnel/stud replacement.

##### Why another wrapper

stud is working perfectly in SSL/TLS based environments, but sometimes all you need is a encryption tunnel without the certification settings and save some time from SSL handshake, so we build `qtunnel` to fit the special network requirement in China. It serves 10m+ connections per day on getqujing.net for months.

### Requirements

qtunnel is writen in [golang 1.3.1](http://golang.org/dl/), after build it can run on every OS now.

### Build

To build `qtunnel`

`$ make`

To test `qtunnel`

`$ make test`

### Usage

	$ ./bin/qtunnel -h
	Usage of ./bin/qtunnel:
		-backend="127.0.0.1:6400": host:port of the backend
		-clientmode=false: if running at client mode
		-crypto="rc4": encryption method
		-listen=":9001": host:port qtunnel listen on
		-logto="stdout": stdout or syslog
		-secret="secret": password used to encrypt the data
 		
`qtunnel` support two crypto method for now: `rc4` and `aes256cfb`, both side of the tunnel should use same `crypto` and same `secret`.



debian/changelog
qtunnel (0.1.0-1) unstable; urgency=low

  * Initial release

 -- Zhou Rui <zhourui@getqujing.com>  Sat, 02 Aug 2014 12:33:02 +0800


src/tunnel/conn_test.go
package tunnel

import (
    "testing"
    "net"
    "bytes"
)

func TestWrite(t *testing.T) {
    done := make(chan bool)
    secret := []byte("secret")
    data := []byte{0, 1, 2, 3}
    ln, err := net.Listen("tcp", "127.1:9444")
    if err != nil {
        t.Error(err)
    }
    defer ln.Close()
    go func() {
        conn, err := ln.Accept()
        if err != nil {
            t.Error(err)
        }
        defer conn.Close()
        cipher := NewCipher("rc4", secret)
        buf := make([]byte, len(data))
        conn.Read(buf)
        cipherData := make([]byte, len(data))
        cipher.encrypt(cipherData, data)
        if (bytes.Compare(cipherData, buf) != 0) {
            t.Fail()
        }
        close(done)
    }()
    conn, err := net.Dial("tcp", "127.1:9444")
    if err != nil {
        t.Error(err)
    }
    conn2 := NewConn(conn, NewCipher("rc4", secret))
    defer conn2.Close()
    conn2.Write(data)
    <-done
}

func TestRead(t *testing.T) {
    done := make(chan bool)
    secret := []byte("secret")
    data := []byte{0, 1, 2, 3}
    ln, err := net.Listen("tcp", "127.1:9444")
    if err != nil {
        t.Error(err)
    }
    defer ln.Close()
    go func() {
        conn, err := ln.Accept()
        if err != nil {
            t.Error(err)
        }
        defer conn.Close()
        cipher := NewCipher("rc4", secret)
        cipherData := make([]byte, len(data))
        cipher.encrypt(cipherData, data)
        conn.Write(cipherData)
        close(done)
    }()
    conn, err := net.Dial("tcp", "127.1:9444")
    if err != nil {
        t.Error(err)
    }
    conn2 := NewConn(conn, NewCipher("rc4", secret))
    defer conn2.Close()
    buf := make([]byte, len(data))
    conn2.Read(buf)
    if (bytes.Compare(data, buf) != 0) {
        t.Fail()
    }
    <-done
}


src/tunnel/cipher_test.go
package tunnel

import (
    "testing"
)

func TestRC4(t *testing.T) {
    secret := []byte("testsecret")
    clearText := "thisISaCLEARtext"
    c := NewCipher("rc4", secret)
    c2 := NewCipher("rc4", secret)
    dst := make([]byte, len(clearText))
    dst2 := make([]byte, len(clearText))
    c.encrypt(dst, []byte(clearText))
    c2.decrypt(dst2, dst)
    if (clearText != string(dst2)) {
        t.Error(string(dst2))
    }
}


src/tunnel/tunnel_test.go
package tunnel

import (
    "testing"
    "net"
    "bytes"
    "time"
)

func TestTunnel(t *testing.T) {
    done := make(chan bool)
    data := []byte{1, 2, 3, 4}
    go backendServer(data, done, t)
    b := NewTunnel("127.1:9446", "127.1:9445", false, "rc4", "secret")
    f := NewTunnel("127.1:9447", "127.1:9446", true, "rc4", "secret")
    go b.Start()
    go f.Start()
    // sleep to wait all servers start
    time.Sleep(100 * time.Millisecond)
    conn, err := net.Dial("tcp", "127.1:9447")
    if err != nil {
        t.Error(err)
    }
    defer conn.Close()
    conn.Write(data)
    // wait for transmission complete
    time.Sleep(100 * time.Millisecond)
    close(done)
}

func backendServer(data []byte, done chan bool, t *testing.T) {
    ln, err := net.Listen("tcp", "127.1:9445")
    if err != nil {
        t.Error(err)
    }
    defer ln.Close()
    conn, err := ln.Accept()
    if err != nil {
        t.Error(err)
    }
    defer conn.Close()
    buf := make([]byte, len(data))
    conn.Read(buf)
    if (bytes.Compare(buf, data) != 0) {
        t.Fail()
    }
    <-done
}


src/tunnel/conn.go
package tunnel

import (
    "net"
)

type Conn struct {
    conn net.Conn
    cipher *Cipher
}

func NewConn(conn net.Conn, cipher *Cipher) *Conn {
    return &Conn{conn, cipher}
}

func (c *Conn) Read(b []byte) (int, error) {
    if c.cipher == nil {
        return c.conn.Read(b)
    }
    cipherData := make([]byte, len(b))
    n, err := c.conn.Read(cipherData)
    if n > 0 {
        c.cipher.decrypt(b[0:n], cipherData[0:n])
    }
    return n, err
}

func (c *Conn) Write(b []byte) (int, error) {
    if c.cipher == nil {
        return c.conn.Write(b)
    }
    cipherData := make([]byte, len(b))
    c.cipher.encrypt(cipherData, b)
    return c.conn.Write(cipherData)
}

func (c *Conn) Close() {
    c.conn.Close()
}

func (c *Conn) CloseRead() {
    if conn, ok := c.conn.(*net.TCPConn); ok {
        conn.CloseRead()
    }
}

func (c *Conn) CloseWrite() {
    if conn, ok := c.conn.(*net.TCPConn); ok {
        conn.CloseWrite()
    }
}


src/tunnel/cipher.go
package tunnel

import (
    "log"
    "crypto/rc4"
    "crypto/sha256"
    "crypto/cipher"
)

type Cipher struct {
    enc cipher.Stream
    dec cipher.Stream
}

type chiperCreator func(key []byte) (*Cipher, error)

var cipherMap = map[string]chiperCreator {
    "rc4": newRC4Cipher,
}

func secretToKey(secret []byte) []byte {
    h := sha256.New()
    return h.Sum(secret)
}

func newRC4Cipher(key []byte) (*Cipher, error) {
    c, err := rc4.NewCipher(key)
    if err != nil {
        return nil, err
    }
    c2 := *c

    return &Cipher{c, &c2}, nil
}

func NewCipher(cryptoMethod string, secret []byte) *Cipher {
    cc := cipherMap[cryptoMethod]
    if cc == nil {
        log.Fatalf("unsupported crypto method %s", cryptoMethod)
    }
    c, err := cc(secretToKey(secret))
    if err != nil {
        log.Fatal(err)
    }
    return c
}

func (c *Cipher) encrypt(dst, src []byte) {
    c.enc.XORKeyStream(dst, src)
}

func (c *Cipher) decrypt(dst, src []byte) {
    c.dec.XORKeyStream(dst, src)
}


src/tunnel/tunnel.go
package tunnel

import (
    "io"
    "net"
    "log"
    "time"
    "sync/atomic"
)

type Tunnel struct {
    faddr, baddr *net.TCPAddr
    clientMode bool
    cryptoMethod string
    secret []byte
    sessionsCount int32
}

func NewTunnel(faddr, baddr string, clientMode bool, cryptoMethod, secret string) *Tunnel {
    a1, err := net.ResolveTCPAddr("tcp", faddr)
    if err != nil {
        log.Fatalln("resolve frontend error:", err)
    }
    a2, err := net.ResolveTCPAddr("tcp", baddr)
    if err != nil {
        log.Fatalln("resolve backend error:", err)
    }
    return &Tunnel{
        faddr: a1,
        baddr: a2,
        clientMode: clientMode,
        cryptoMethod: cryptoMethod,
        secret: []byte(secret),
        sessionsCount: 0,
    }
}

func (t *Tunnel) pipe(dst, src *Conn, c chan int64) {
    defer func() {
        dst.CloseWrite()
        src.CloseRead()
    }()
    n, err := io.Copy(dst, src)
    if err != nil {
        log.Print(err)
    }
    c <- n
}

func (t *Tunnel) transport(conn net.Conn) {
    start := time.Now()
    conn2, err := net.DialTCP("tcp", nil, t.baddr)
    if err != nil {
        log.Print(err)
        return
    }
    connectTime := time.Now().Sub(start)
    start = time.Now()
    cipher := NewCipher(t.cryptoMethod, t.secret)
    readChan := make(chan int64)
    writeChan := make(chan int64)
    var readBytes, writeBytes int64
    atomic.AddInt32(&t.sessionsCount, 1)
    var bconn, fconn *Conn
    if t.clientMode {
        fconn = NewConn(conn, nil)
        bconn = NewConn(conn2, cipher)
    } else {
        fconn = NewConn(conn, cipher)
        bconn = NewConn(conn2, nil)
    }
    go t.pipe(bconn, fconn, writeChan)
    go t.pipe(fconn, bconn, readChan)
    select {
    case readBytes = <-readChan:
        writeBytes = <-writeChan
    case writeBytes = <-writeChan:
        readBytes = <-readChan
    }
    transferTime := time.Now().Sub(start)
    log.Printf("r:%d w:%d ct:%.3f t:%.3f [#%d]", readBytes, writeBytes,
        connectTime.Seconds(), transferTime.Seconds(), t.sessionsCount)
    atomic.AddInt32(&t.sessionsCount, -1)
}

func (t *Tunnel) Start() {
    ln, err := net.ListenTCP("tcp", t.faddr)
    if err != nil {
        log.Fatal(err)
    }
    defer ln.Close()

    for {
        conn, err := ln.AcceptTCP()
        if err != nil {
            log.Println("accept:", err)
            continue
        }
        go t.transport(conn)
    }
}


src/qtunnel/main.go
package main

import (
    "os"
    "os/signal"
    "syscall"
    "log"
    "log/syslog"
    "flag"
    "tunnel"
)

func waitSignal() {
    var sigChan = make(chan os.Signal, 1)
    signal.Notify(sigChan)
    for sig := range sigChan {
        if sig == syscall.SIGINT || sig == syscall.SIGTERM {
            log.Printf("terminated by signal %v\n", sig)
            return
        } else {
            log.Printf("received signal: %v, ignore\n", sig)
        }
    }
}

func main() {
    var faddr, baddr, cryptoMethod, secret, logTo string
    var clientMode bool
    flag.StringVar(&logTo, "logto", "stdout", "stdout or syslog")
    flag.StringVar(&faddr, "listen", ":9001", "host:port qtunnel listen on")
    flag.StringVar(&baddr, "backend", "127.0.0.1:6400", "host:port of the backend")
    flag.StringVar(&cryptoMethod, "crypto", "rc4", "encryption method")
    flag.StringVar(&secret, "secret", "secret", "password used to encrypt the data")
    flag.BoolVar(&clientMode, "clientmode", false, "if running at client mode")
    flag.Parse()

    log.SetOutput(os.Stdout)
    if logTo == "syslog" {
        w, err := syslog.New(syslog.LOG_INFO, "qtunnel")
        if err != nil {
            log.Fatal(err)
        }
        log.SetOutput(w)
    }

    t := tunnel.NewTunnel(faddr, baddr, clientMode, cryptoMethod, secret)
    log.Println("qtunnel started.")
    go t.Start()
    waitSignal()
}
